File: client/src/modules/nexotype/schemas/knowledge_graph/variant-phenotype.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const VariantPhenotypeSchema = z.object({
  id: z.number(),
  variant_id: z.number().int(),
  phenotype_id: z.number().int(),
  effect_size: z.string().max(50).nullable().optional(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateVariantPhenotypeSchema = z.object({
  variant_id: z.number().int(),
  phenotype_id: z.number().int(),
  effect_size: z.string().max(50).nullable().optional(),
});

export const UpdateVariantPhenotypeSchema = z.object({
  variant_id: z.number().int().optional(),
  phenotype_id: z.number().int().optional(),
  effect_size: z.string().max(50).nullable().optional(),
});

// ==========================================
// Types
// ==========================================

export type VariantPhenotype = z.infer<typeof VariantPhenotypeSchema>;
export type CreateVariantPhenotype = z.infer<typeof CreateVariantPhenotypeSchema>;
export type UpdateVariantPhenotype = z.infer<typeof UpdateVariantPhenotypeSchema>;

// ==========================================
// Response Types
// ==========================================

export type VariantPhenotypeResponse = {
  success: boolean;
  data?: VariantPhenotype;
  error?: string;
};

export type VariantPhenotypesResponse = {
  success: boolean;
  data?: VariantPhenotype[];
  count?: number;
  error?: string;
};

File: server/apps/nexotype/schemas/knowledge_graph/variant_phenotype_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class VariantPhenotypeCreate(BaseModel):
    """Schema for creating a new variant phenotype"""
    variant_id: int
    phenotype_id: int
    effect_size: str | None = None


class VariantPhenotypeUpdate(BaseModel):
    """Schema for updating a variant phenotype"""
    variant_id: int = None
    phenotype_id: int = None
    effect_size: str | None = None


# ==========================================
# Response Schemas
# ==========================================

class VariantPhenotypeDetail(BaseModel):
    """Schema for variant phenotype details"""
    id: int
    variant_id: int
    phenotype_id: int
    effect_size: str | None = None
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class VariantPhenotypeListResponse(BaseModel):
    """Response schema for listing variant phenotypes"""
    success: bool
    data: list[VariantPhenotypeDetail] | None = None
    count: int = 0
    error: str | None = None


class VariantPhenotypeResponse(BaseModel):
    """Response schema for single variant phenotype operations"""
    success: bool
    data: VariantPhenotypeDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None
